webhooks/pkg/webhooks/manager/event.go

```diff
@@ -0,0 +1,284 @@
+	AdmissionReview *v1beta1.AdmissionReview
+	crd             resources.GenericResource
+	oldCrd          resources.GenericResource
+}
+
+// NewEvent is a constructor for Event
+func NewEvent(ar *v1beta1.AdmissionReview) (*Event, error) {
+	arjson, err := json.Marshal(ar)
+	req := ar.Request
+	glog.Infof(`
+    Creating event for
+	%v
+	Namespace=%v
+	Request Name=%v
+	UID=%v
+	patchOperation=%v
+	UserInfo=%v`,
+		req.Kind,
+		req.Namespace,
+		req.Name,
+		req.UID,
+		req.Operation,
+		req.UserInfo)
+	crd, err := resources.GetGenericResource(ar.Request.Object.Raw)
+	glog.Infof("Resource name : %v", crd.Name)
+	if err != nil {
+		glog.Errorf("Admission review JSON: %v", string(arjson))
+		glog.Errorf("Could not get the GenericResource object %v", err)
+		return nil, err
+	}
+
+	var oldCrd resources.GenericResource = resources.GenericResource{}
+	if len(ar.Request.OldObject.Raw) != 0 {
+		oldCrd, err = resources.GetGenericResource(ar.Request.OldObject.Raw)
+		if err != nil {
+			glog.Errorf("Admission review JSON: %v", string(arjson))
+			glog.Errorf("Could not get the old GenericResource object %v", err)
+			return nil, err
+		}
+	}
+	return &Event{
+		AdmissionReview: ar,
+		crd:             crd,
+		oldCrd:          oldCrd,
+	}, nil
+}
+
+func (e *Event) isStateChanged() bool {
+	glog.Infof("Checking state change new state: %s\n", e.crd.Status.State)
+	glog.Infof("Checking state change old state: %s\n", e.oldCrd.Status.State)
+	return e.crd.Status.State != e.oldCrd.Status.State
+}
+
+func (e *Event) isDeleteTriggered() bool {
+	return e.crd.Status.State == Delete
+}
+
+func (e *Event) isPlanChanged() (bool, error) {
+	appliedOptionsNew, err := e.crd.GetAppliedOptions()
+	if err != nil {
+		return false, err
+	}
+	appliedOptionsOld, err := e.oldCrd.GetAppliedOptions()
+	if err != nil {
+		return false, err
+	}
+	return appliedOptionsNew.PlanID != appliedOptionsOld.PlanID, nil
+}
+
+func (e *Event) isCreate() (bool, error) {
+	lo, err := e.crd.GetLastOperation()
+	if err != nil {
+		return false, err
+	}
+	return lo.Type == loCreate, nil
+}
+
+func (e *Event) isUpdate() (bool, error) {
+	lo, err := e.crd.GetLastOperation()
+	if err != nil {
+		return false, err
+	}
+	return lo.Type == loUpdate, nil
+}
+
+func (e *Event) isSucceeded() bool {
+	return e.crd.Status.State == Succeeded
+}
+
+func (e *Event) isDirector() bool {
+	return e.crd.Kind == Director
+}
+
+func (e *Event) isDocker() bool {
+	return e.crd.Kind == Docker
+}
+
+func (e *Event) isMeteringEvent() (bool, error) {
+	// An event is metering event if
+	// Create succeeded
+	// or Update Succeeded
+	// or Delete Triggered
+	isUpdate, err := e.isUpdate()
+	if err != nil {
+		return false, err
+	}
+	isCreate, err := e.isCreate()
+	if err != nil {
+		return false, err
+	}
+	isPlanChanged, err := e.isPlanChanged()
+	if err != nil {
+		return false, err
+	}
+	if e.isDirector() && e.isStateChanged() {
+		if e.isSucceeded() {
+			return (isUpdate && isPlanChanged) || isCreate, nil
```

Comment:
Since we have seen many cases, where last_operation type / appliedOptions is not populated in error cases, maybe it is good idea to move init of ```isUpdate```, ```isCreate``` and ```isPlanChanged``` variables here after ```isSucceeded``` check.